### PR TITLE
[WIP] Reach ID TypeScript Client

### PR DESCRIPTION
**Pull request by chrisnevers**
No tests yet

### CODE REVIEW

File: hs/hie.yaml [hunk 1]
```diff
@@ -2,6 +2,10 @@ cradle:
   stack:
     - path: "./src"
       component: "reach:lib"
+    - path: "./reach-everest/hs/src"
+      component: "reach:lib"
+    - path: "./reach-everest/hs/app-closed/reach-idd"
+      component: "reach:exe:reach-idd"
```
Comment (jeapostrophe) [inline hs/hie.yaml:8]: If this messes up hie when you don't have Everest, we may need to put this in the Everest linking script

File: hs/src/Reach/AST/ID/Req.hs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
```
Comment (jeapostrophe) [inline hs/src/Reach/AST/ID/Req.hs:1]: I don't think this is an AST. I think this stuff goes in `hs/src/Reach/ID/{Req,Client}.hs`

File: hs/src/Reach/AST/ID/Req.hs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  , phoneNumber :: String
+  , mfaSms      :: Maybe String
+  , mfaEmail    :: Maybe String
+  , mfaTotp     :: Maybe Bool
+ } deriving (Eq, Ord, Show, Generic)
+
+instance ToJSON User
+instance FromJSON User
+instance ToJWT User
+instance FromJWT User
```
Comment (jeapostrophe) [inline hs/src/Reach/AST/ID/Req.hs:90]: I assume that the server will use this module too?

File: hs/src/Reach/IdClient.hs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+import           Control.Monad.Reader
+import           Data.Aeson
+import           Data.ByteString        (ByteString)
+import           Data.IORef
+import           Data.Maybe
+import           Network.HTTP.Client
+import           Network.HTTP.Simple
+import           Reach.AST.ID.Req
+import           Reach.Util
+
```
Comment (jeapostrophe) [inline hs/src/Reach/IdClient.hs:30]: Heresy!

File: hs/src/Reach/IdClient.hs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    liftIO . flip writeIORef (Just a) =<< asks f
+  return a
+
+saveMfaInfo :: MfaInfo -> App MfaInfo
+saveMfaInfo = save e_mfaInfo
+
+saveUser :: User -> App User
+saveUser = save e_user
+
+
```
Comment (jeapostrophe) [inline hs/src/Reach/IdClient.hs:141]: Yuck

File: hs/src/Reach/IdClient.hs [hunk 6]
```diff
@@ -0,0 +1,207 @@
+module Reach.IdClient
```
Comment (jeapostrophe) [inline hs/src/Reach/IdClient.hs:1]: This code isn't actually used anywhere so it is confusing to interpret for me. I think you need to add a trivial command line tool that uses this library and that you write shell script tests over in the server code base. (That's an idea, not a command)

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: hs/src/Reach/AST/ID/Req.hs (hunk 3 above)
Comment (chrisnevers) [reply hs/src/Reach/AST/ID/Req.hs:90]: yup

File: hs/src/Reach/IdClient.hs (hunk 4 above)
Comment (chrisnevers) [reply hs/src/Reach/IdClient.hs:30]: you need a better editor to do this formatting for you 😛

File: hs/src/Reach/IdClient.hs (hunk 6 above)
Comment (chrisnevers) [reply hs/src/Reach/IdClient.hs:1]: sounds good, I'll work on that

**Review by jeapostrophe — APPROVED** [review]

**Comment by jeapostrophe** [comment]
Can merge after conflict resolved

**Review by jeapostrophe — APPROVED** [review]
Merge after CI

**Pull request merged**